src/widget/button.rs

```diff
@@ -96,18 +101,20 @@ impl<T: Data> Widget<T> for Label {
         None
     }
 
-    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: Option<&T>, _data: &T, _env: &Env) {}
+    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: Option<&T>, data: &T, env: &Env) {
+        self.text.display_text(data, env);
```

Comment:
What I'd *really* like to see here is that your resolver returns a boolean indicating whether the resolved string has changed, and if so, invalidate. As it is, I think you're underinvalidating here, as the text could change and it won't ask for a repaint.

I think we need to implement at least a basic version of finer-grained invalidation before too long, otherwise there's no real motivation to get invalidation right.

Reply:
Okay, i've made this change, seems clean.